### PR TITLE
[runtime] Fixes threadpool tests not to rely on arbitrary sleep timeout

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

**Comment by akoeplinger** [comment]
@monojenkins build failed

File: mono/tests/Makefile.am [hunk 1]
```diff
@@ -2866,7 +2864,9 @@ TESTS_UNHANDLED_EXCEPTION_255_SRC =	\
 	unhandled-exception-5.cs	\
 	unhandled-exception-6.cs	\
 	unhandled-exception-7.cs	\
-	unhandled-exception-8.cs
+	unhandled-exception-8.cs	\
+	threadpool-exceptions2.cs	\
```
Comment (luhenry) [inline mono/tests/Makefile.am]: This means the `threadpool-exceptions2.cs` and `threadpool-exceptions5.cs` tests are expected to have exit code = 255? That sounds wrong.

File: mono/tests/threadpool-exceptions2.cs [hunk 2]
```diff
@@ -2,38 +2,46 @@
...
+			return;
+		}
+
+		if (!e.IsTerminating) {
+			Environment.Exit (4);
+			return;
 		}
-		Environment.Exit (return_value);
+
+		mre.Set ();
```
Comment (luhenry) [inline mono/tests/threadpool-exceptions2.cs]: Why not having the same `Environment.Exit (0)` as the other tests?

File: mono/tests/threadpool-exceptions5.cs [hunk 3]
```diff
@@ -2,37 +2,47 @@
...
-		lock (monitor) {
-			Monitor.Wait (monitor);
-		}
-		Thread.Sleep (1000);
-		return 1;
+
+		if (!mre.WaitOne (10000))
+			return 2;
+
+		GC.Collect ();
```
Comment (luhenry) [inline mono/tests/threadpool-exceptions5.cs]: What's the need for these 2 new function calls?

File: mono/tests/threadpool-exceptions2.cs [hunk 4]
```diff
@@ -2,38 +2,46 @@
...
-		lock (monitor) {
-			Monitor.Wait (monitor);
-		}
-		Thread.Sleep (1000);
-		return 1;
+
+		if (!mre.WaitOne (10000))
+			return 2;
+
+		GC.Collect ();
```
Comment (luhenry) [inline mono/tests/threadpool-exceptions2.cs]: What's the need for these 2 new function calls?

File: mono/tests/Makefile.am (hunk 1 above)
Comment (marek-safar) [reply mono/tests/Makefile.am]: Yes, that's expected. They are testing the unhandled process exit caused by user exception

File: mono/tests/threadpool-exceptions2.cs (hunk 4 above)
Comment (marek-safar) [reply mono/tests/threadpool-exceptions2.cs]: They are not strictly necessary, could remove them but don't hurt either but it make the repro more reliable

File: mono/tests/threadpool-exceptions2.cs (hunk 2 above)
Comment (marek-safar) [reply mono/tests/threadpool-exceptions2.cs]: Because this is exit path for unhandled exception

File: mono/tests/threadpool-exceptions5.cs (hunk 3 above)
Comment (marek-safar) [reply mono/tests/threadpool-exceptions5.cs]: same as above, could remove then but they help to ensure there is no other side effect

**Pull request merged**